File: client/tester.js
const https = require('https');
const fs = require('fs');
const crypto = require('crypto');
const { URL } = require('url');

const username = "<YOUR_USERNAME>";
const key_path = "../keys/<YOUR_PRIVATE_KEY_FILE_NAME>";

const test_file_path = "./filtered_variables1_formatted.csv"; // Try swapping out 1 with 2, 3 or 4 and see the output change!

const url = new URL("https://bioage.ambr.no");

// change these for different outputs
const gender = "female";
const language = "en";

const challenge_path = "/auth/challenge";
const verify_path = "/auth/verify";
const secure_path = "/auth/secure";
const variable_path = "/bioage_variable_metadata"
const quest_path = "/questionnaire?gender="+gender+"&language="+language
const bioage_path = "/bioage_predictor?gender="+gender;

function makeRequest(method,payload,custom_options){
    return new Promise((resolve,reject) => {
        const options = {
            ...custom_options,
            host: url.host,
            port: url.port,
            method,
        }
        let result = "";
        let request = https.request(options, function(response) {
            response.on('data', function (chunk) {
                result = result+chunk;
            });
            response.on('end', function(){
                resolve(JSON.parse(result))
            });
        });
        request.on('error', function(err) {
            reject(err);
        });
        request.write(payload);
        request.end()
    })
}

// Get variables names and descriptions with token
function getVarsWithAuth(token) {
    const var_options = {
        path: variable_path,
        headers: {
            'Authorization': "Bearer "+token
        }
    }
    makeRequest('GET',"",var_options)
    .then((response) => {
        // console.log(response.variables); // Takes a lot of space in the output, but please uncomment for a better display than the file
        fs.writeFileSync('./variables_with_desc.txt', JSON.stringify(response.variables));
        console.log("Variables written to file");
    },
        (err) => {
            console.log(err);
    });
}

// Get questionnaire json
function getQuestJsonWithAuth(token) {
    const quest_options = {
        path: quest_path,
        headers: {
            'Authorization': "Bearer "+token
        }
    }
    makeRequest('GET',"", quest_options)
    .then((response) => {
        console.log(typeof response);
        fs.writeFileSync('./questionnaire_json.json', JSON.stringify(response));
        console.log("Questionnaire written to file");
    },
        (err) => {
            console.log(err);
    });
}

// Bioage call with token
function bioAgeWithAuth(token) {
    const bio_options = {
        path: bioage_path,
        headers: {
            'Content-Type': 'text/csv',
            'Authorization': "Bearer "+token
        }
    }

    fs.readFile(test_file_path, 'utf-8', function (err, data) {
        if (err) {
            console.log("Error reading the file: " + err);
            process.exit(-2);
        }
        if(data) {
            makeRequest('POST',data,bio_options)
            .then((bioage) => {
                console.log("User: "+bioage.eid);
                console.log("Bioage: "+bioage.bioage);
                console.log("with metadata: "+bioage.metadata); // Metadata will contain many exciting things in the future ~
            },
            (err) => {
                console.log(err);
            });
        }
        else {
            console.log("No data");
            process.exit(-1);
        }
    });
}

// Call the API
makeRequest('GET',"",{}).then(
    (response) => {
        console.log(response);
    }
)

// ------------------- Authentication Guide ---------------------------

// 1. Load private key
const private_pem_key = fs.readFileSync(key_path, 'utf-8')
const private_key = crypto.createPrivateKey(
    {
        key: private_pem_key,
        format: 'pem',
    }
)

// 2. Request challenge (nonce) from server
const challenge_options = {
    path: challenge_path
}

makeRequest(
    'POST',
    '{"username":"'+username+'"}',
    challenge_options
).then(
    (nonceResponse) => {
        console.log("Nonce received: "+nonceResponse.nonce);

// 3. Sign the nonce
        const signature = crypto.sign("sha256" /*must match server*/, Buffer.from(nonceResponse.nonce,'utf-8'), {
        key: private_key,
        padding: crypto.constants.RSA_PKCS1_PADDING, 
        });
        const encoded_signature = signature.toString('base64');

// 4. Send signature to /auth/verify
        let verify_options = {
            path: verify_path
        }
        makeRequest(
            'POST',
            '{"username": "'+username+'","signature": "'+encoded_signature+'"}',
            verify_options
        )
        .then(
            (tokenResponse) => {
                console.log("Server response: "+JSON.stringify(tokenResponse));
                
// 5. Secure authentication
                let token = tokenResponse.token
                const secure_options = {
                path: secure_path,
                headers: {
                    'Authorization': "Bearer "+token
                }
            }

            makeRequest('POST',"",secure_options).then(
                (secureResponse) => {
                    console.log("Secure response: ", secureResponse);

                    // Once you have a valid token, you can make a call for bioage, variable names and the questionnaire
                    getVarsWithAuth(token);

                    getQuestJsonWithAuth(token); // response in questionnaire_json.json

                    bioAgeWithAuth(token); // response in variables_with_desc.txt

                }
            )

            },
            (err) => {
            console.log(err);
            }
        )
    },
    (err) => {
        console.log(err);
    }
)

File: client/tester.py
import base64
import httpx
import pandas as pd
import time
from cryptography.hazmat.primitives import serialization, hashes
from cryptography.hazmat.primitives.asymmetric import padding

USERNAME = "<YOUR_USERNAME>"
KEY_PATH = "../keys/<YOUR_PRIVATE_KEY_FILE_NAME>"
API_BASE = "https://bioage.ambr.no"

# 1. Authenticate towards server and store token client side
# 1.1. Load private key
print("1. Authenticate towards server")
with open(KEY_PATH, "rb") as f:
    private_key = serialization.load_pem_private_key(f.read(), password=None)

# 1.2. Request challenge (nonce) from server
challenge_url = f"{API_BASE}/auth/challenge"
res = httpx.post(challenge_url, json={"username": USERNAME})
if res.status_code != 200:
    raise SystemExit(f"Failed to get challenge: {res.text}")

nonce = res.json()["nonce"]
print(f"  Nonce received: {nonce}")

# 1.3. Sign the nonce
signature = private_key.sign(
    nonce.encode(),
    padding.PKCS1v15(),
    hashes.SHA256()  # must match the server
)
encoded_signature = base64.b64encode(signature).decode()

# 1.4. Send signature to /auth/verify
verify_url = f"{API_BASE}/auth/verify"
verify_payload = {
    "username": USERNAME,
    "signature": encoded_signature
}

res = httpx.post(verify_url, json=verify_payload)
token = res.json()["token"]
print(f"  Server response: {res.status_code} {res.text}")

# 1.5 Secure everything went smoothly
secure_url = f"{API_BASE}/auth/secure"
secure_res = httpx.post(
    secure_url,
    headers={"Authorization": f"Bearer {token}"}
)
print("  Secure response:", secure_res.json())

# 2. Test bioage endpoints
# 2.1. Predict bioage from csv file input - female and male
print("2. Testing bioage endpoints")
print("  Securely predicting bioage from csv input...")
f = open("./filtered_variables1_formatted.csv", "r")
lines = f.read()

tic = time.time()
predictor_url = f"{API_BASE}/bioage_predictor?gender=female"
secure_res = httpx.post(
    predictor_url,
    content=lines,
    headers={
        "Content-Type": "text/csv",
        "Authorization": f"Bearer {token}"
    }
)
print(f"    Bioage prediction response from csv input in {time.time() - tic:.2f}s, female:", secure_res.json())

tic = time.time()
predictor_url = f"{API_BASE}/bioage_predictor?gender=male"
secure_res = httpx.post(
    predictor_url,
    content=lines,
    headers={
        "Content-Type": "text/csv",
        "Authorization": f"Bearer {token}"
    }
)
print(f"    Bioage prediction response from csv input in {time.time() - tic:.2f}s, male:", secure_res.json())


# 2.2. Predict bioage from json input - female and male
print("  Securely predicting bioage from json input...")
df = pd.read_csv("./filtered_variables1_formatted.csv")
input_data = df.to_dict(orient="records")[0]

tic = time.time()
predictor_url = f"{API_BASE}/bioage_predictor?gender=female"
secure_res = httpx.post(
    predictor_url,
    json=input_data,
    headers={
        "Content-Type": "application/json",
        "Authorization": f"Bearer {token}"
    }
)
print(f"    Bioage prediction response from json input in {time.time() - tic:.2f}s, female:", secure_res.json())

tic = time.time()
predictor_url = f"{API_BASE}/bioage_predictor?gender=male"
secure_res = httpx.post(
    predictor_url,
    json=input_data,
    headers={
        "Content-Type": "application/json",
        "Authorization": f"Bearer {token}"
    }
)
print(f"    Bioage prediction response from json input in {time.time() - tic:.2f}s, male:", secure_res.json())


# 2.3. Get bioage variables metadata
print("  Securely fetching bioage variable metadata...")
variable_metadata_url = f"{API_BASE}/bioage_variable_metadata"
secure_res = httpx.get(
    variable_metadata_url,    
    headers={
        "Content-Type": "application/json",
        "Authorization": f"Bearer {token}"
    }
)
variable_metadata = secure_res.json()
print(f"    Variable metadata response contains {(variable_metadata['n_variables'])} variables")


# 3. Test questionnaire endpoints
# 3.1. English questionnaire, female
print("3. Testing questionnaire endpoints")
print("  Securely fetching English questionnaire for female...")
questionnaire_url = f"{API_BASE}/questionnaire"
questionnaire_params = {"language": "en", "gender": "f"}
secure_res = httpx.get(
    questionnaire_url,
    params=questionnaire_params,
    headers={
        "Content-Type": "application/json",
        "Authorization": f"Bearer {token}"
    }
)
questionnaire_data = secure_res.json()
print(f"    Questionnaire response contains {len(questionnaire_data['response'])} questions")


# 3.2. Norwegian questionnaire, female
print("  Securely fetching Norwegian questionnaire for female...")
questionnaire_url = f"{API_BASE}/questionnaire"
questionnaire_params = {"language": "no", "gender": "f"}
secure_res = httpx.get(
    questionnaire_url,
    params=questionnaire_params,
    headers={
        "Content-Type": "application/json",
        "Authorization": f"Bearer {token}"
    }
)
questionnaire_data = secure_res.json()
print(f"    Questionnaire response contains {len(questionnaire_data['response'])} questions")



# 3.3. English questionnaire, male
print("  Securely fetching English questionnaire for male...")
questionnaire_url = f"{API_BASE}/questionnaire"
questionnaire_params = {"language": "en", "gender": "m"}
secure_res = httpx.get(
    questionnaire_url,
    params=questionnaire_params,
    headers={
        "Content-Type": "application/json",
        "Authorization": f"Bearer {token}"
    }
)
questionnaire_data = secure_res.json()
print(f"    Questionnaire response contains {len(questionnaire_data['response'])} questions")


# 3.4. Norwegian questionnaire, male
print("  Securely fetching Norwegian questionnaire for male...")
questionnaire_url = f"{API_BASE}/questionnaire"
questionnaire_params = {"language": "no", "gender": "m"}
secure_res = httpx.get(
    questionnaire_url,
    params=questionnaire_params,
    headers={
        "Content-Type": "application/json",
        "Authorization": f"Bearer {token}"
    }
)
questionnaire_data = secure_res.json()
print(f"    Questionnaire response contains {len(questionnaire_data['response'])} questions")